keps/sig-storage/2644-honor-pv-reclaim-policy/README.md

```diff
@@ -0,0 +1,380 @@
+```go
+// shouldDelete returns whether a volume should have its backing volume
+// deleted, i.e. whether a Delete is "desired"
+func (ctrl *ProvisionController) shouldDelete(ctx context.Context, volume *v1.PersistentVolume) bool {
+	if deletionGuard, ok := ctrl.provisioner.(DeletionGuard); ok {
+		if !deletionGuard.ShouldDelete(ctx, volume) {
+			return false
+		}
+	}
+
+	if ctrl.addFinalizer && !ctrl.checkFinalizer(volume, finalizerPV) && volume.ObjectMeta.DeletionTimestamp != nil {
+		return false
+	} else if volume.ObjectMeta.DeletionTimestamp != nil { //<========== DeletionTimestamp is set on the PV since it's deleted first.
+		return false
+	}
+    // Remaining code below skipped for brevity...
+    // Please refer sig-storage-lib-external-provisioner/controller/controller.go for the full code.
+```
+
+The main approach in fixing the issue involves introducing new finalizer `kubernetes.io/pv-deletion-protection` on the Persistent Volume when the controller sees it. Adding the finalizer prevents the PV to be removed from the API server. The finalizer will be removed only after the physical volume on the storage system is deleted.
+
+When the PVC is deleted, the PV is moved into `Released` state and following checks are made:
+
+1. Plugin:
+If the volume has the finalizer `kubernetes.io/pv-deletion-protection`, then, `DeletionTimestamp` checks can be ignored.
+
+3. CSI driver
+
+If when processing the PV update it is observed that it has `kubernetes.io/pv-deletion-protection` finalizer and
+`DeletionTimestamp` set, then the volume deletion request is forwarded to the driver(provided other pre-defined 
+conditions are met).
+
+
+### Risks and Mitigations
+
+The primary risk is volume deletion to a user that expect the current behavior, i.e, they do not expect the volume to be
+deleted from the underlying storage infrastructure when for a bound PV-PVC pair, the PV is deleted followed by deleting
+the PVC. This can be mitigated by initially introducing the fix behind a feature gate, followed by explicitly deprecating
+the current behavior.
+
+## Design Details
+
+A feature gate named `HonorPVReclaimPolicy` will be introduced for both `kube-controller-manager` and `external-provisioner`.
+
+A new finailzer `PVDeletionProtectionFinalizer` will be added in syncVolume() of sig-storage-lib-external-provisioner/controller/controller.go if needed. The controller will find the corresponding StorageClass from the PV and find the provisioner specified in the StorageClass. The new finalizer will be added only if the provisioner in the StorageClass matches the provisioner name of the external-provisioner. This means statically provisioned PVs won't get this finalizer unless if the correct storageClassName is provided in the PV. Also PVs provisioned by in-tree volume plugins won't get this new finalizer. We can add this new finalizer for in-tree volume plugins later if needed in the future.
```

Comment:
Again, refer to existing finalizer and `AddFinalizer()` config option.

Reply:
Fixed.